### PR TITLE
Flatpak support for rpmostree payload

### PR DESCRIPTION
**Pull request by jkonecny12**
This will add logic to install all the flatpak apps and runtimes from the local repository on stage2 path `/flatpak/repo`.

This change set will add flatpak "payload" which contains logic to install flatpaks after rpmostree installation. Also I have to change how installation tasks for rpmostreepayload are registered so the installed flatpaks won't be erased by the prepare mount points installation task in the rpmostree payload.

This solution will be used as a base for creating a proper payload after rpmostree will be a module.

Possible improvements:
* Better error reporting and reaction 
    * Ask user if we can skip the failing flatpak?
* Better progress
    * Report user progress of the current flatpak. Right now it only tells you what flatpak is installed but runtimes could have plenty of gigs so it could take some time before user see a change.

*Resolves: rhbz#1734970*

### CODE REVIEW

File: pyanaconda/payload/flatpak.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        :param operation: object describing the operation
+        :type operation: Flatpak.TransactionOperation instance
+        :param error: object containing error description
+        :type error: GLib.Error instance
+        :param details: information if the error was fatal
+        :type details: int value of Flatpak.TransactionErrorDetails
+        """
+        self._log_operation(operation, "failed")
+        log.error("Flatpak installation failed with message: '%s' -- error is fatal %s",
+                  error.message, self._error_fatal)
```
Comment (rvykydal) [inline pyanaconda/payload/flatpak.py]: Is there _error_fatal attribute in the instance?

File: pyanaconda/payload/rpmostreepayload.py [hunk 2]
```diff
@@ -53,6 +54,10 @@ def __init__(self, *args, **kwargs):
         self._internal_mounts = []
         self._locale_map = None
 
+        self._flatpak_payload = FlatpakPayload(conf.target.system_root)
+        # Initialize temporal repo to enable reading of the remote
+        self._flatpak_payload.initialize_with_path("/var/tmp/anaconda-flatpak-temp")
```
Comment (rvykydal) [inline pyanaconda/payload/rpmostreepayload.py]: Is it safe to have FlatpackPayload (even initialized) in RPMOSTreePayload by default? I mean from the perspective of use cases without flatpacks - now running code (and hitting potential bugs) which is not necessary to run.
But maybe we don't care at this point.
And shouldn't we check if the payload is available before trying to initialize it?

**Comment by jkonecny12** [comment]
jenkins, test this please

**Comment by jkonecny12** [comment]
Jenkins, it's ok to test.

File: pyanaconda/payload/rpmostreepayload.py (hunk 2 above)
Comment (jkonecny12) [reply pyanaconda/payload/rpmostreepayload.py]: Good point maybe it would make sense to promote the `is_available` as a static method.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Fixed pylint errors

File: pyanaconda/payload/flatpak.py (hunk 1 above)
Comment (jkonecny12) [reply pyanaconda/payload/flatpak.py]: A valid bug found also by pylint. I should have run the pylint before posting PR.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Switch `is_available()` method to a class method
* Avoid creating of the flatpak object when repository is not available.

File: pyanaconda/payload/flatpak.py (hunk 1 above)
Comment (jkonecny12) [reply pyanaconda/payload/flatpak.py]: Fixed.

File: pyanaconda/payload/rpmostreepayload.py (hunk 2 above)
Comment (jkonecny12) [reply pyanaconda/payload/rpmostreepayload.py]: Fixed by checking is available before creation.

**Comment by jkonecny12** [comment]
Hi, @cgwalters this is a heads-up if you agree with the work done here.

File: pyanaconda/payload/rpmostreepayload.py [hunk 3]
```diff
@@ -53,6 +54,14 @@ def __init__(self, *args, **kwargs):
         self._internal_mounts = []
         self._locale_map = None
 
+        # flatpak_payload will be None if it can't be used in this environment (don't have repo)
+        self._flatpak_payload = None
+
+        if FlatpakPayload.is_available():
```
Comment (poncovka) [inline pyanaconda/payload/rpmostreepayload.py]: There could be a new type of the `RPMOSTreePayload` with the flatpack support always enabled, so you don't have to check whether `self._flatpak_payload` is None all the time.

Comment (jkonecny12) [reply pyanaconda/payload/rpmostreepayload.py]: Great solution. Working on that!

File: pyanaconda/payload/flatpak.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                  error.message, details == TransactionErrorDetails.FATAL)
+
+    @staticmethod
+    def _log_operation(operation, state):
+        operation_type_str = TransactionOperationType.to_string(operation.get_operation_type())
+        log.debug("Flatpak operation: %s of ref %s state %s",
+                  operation_type_str, operation.get_ref(), state)
+
+
+class BaseRefsList(ABC):
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply pyanaconda/payload/flatpak.py]: Yeah I know but I'm using official names for that.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Split flatpaks logic to a separate rpmostree payload (thanks for suggestion @poncovka!)

File: pyanaconda/payload/flatpak.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/payload/flatpak.py]: That is ok, but it should be explained in the docstring.

File: pyanaconda/installation.py [hunk 5]
```diff
@@ -331,14 +331,11 @@ def run_pre_install():
     installation_queue.append(payload_install)
 
     # for some payloads storage is configured after the payload is installed
-    if not payload.needs_storage_configuration:
+    if not payload.needs_storage_configuration and not conf.target.is_directory:
```
Comment (cgwalters) [inline pyanaconda/installation.py]: This was separated in this way for a reason I think...maybe using lorax/LMC to create live rpm-ostree? I forget.  May not be relevant anymore.  `git annotate` would probably know though.

Comment (jkonecny12) [reply pyanaconda/installation.py]: Ups you are correct. We would probably broke the --dirinstall by calling the part of this code there.
Thanks a lot for your idea!

Comment (jkonecny12) [reply pyanaconda/installation.py]: Also there is a question, would dirinstall without this code be usable. The code creates /var and similar mountpoints so I can't imagine system usable without this. Also is dirinstall even useful for ostree?

What do you think @cgwalters ?

Comment (jkonecny12) [reply pyanaconda/installation.py]: So or so behavior is the same as before so I'm merging this. We can fix this later.

Comment (jkonecny12) [reply pyanaconda/installation.py]: I was able to find out when the check for dir install was removed and it's this commit:
664ef7b43f9102aa9332d0db5b7d13f8ece436f0

It was removed by Brian and the reason was migration to bind mounts. So this seems as a targeted change which improved the situation. In that case I'm taking this as works as expected.

**Pull request merged**